win: port blk_recovery

ChandKV opened the pull request:
- Fix pmempool_info\TEST5.PS1 to not run in nondebug mode as the output
  includes hexdump of pmemblk structure which is different (i.e bigger)
  in debug builds.  And match file is modeled for verifying output from
  the debug binary
- Fix runtest function to set the EXE_DIR based on the build that we
  are testing
- Fix obj_locks project to have right dependencies

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/1057)
<!-- Reviewable:end -->

lplewa commented:
Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/test/blk_recovery/blk_recovery.c, line 182 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOOnZD2ChlTPWYuLdeb:-KOOnZD2ChlTPWYuLdec:b-b4whnz) ([raw file](https://github.com/pmem/nvml/blob/88101937ff7e15d69241b82b7fa4847630e8971f/src/test/blk_recovery/blk_recovery.c#L182)):*
> ```C
> 	}
> #else
> 	} __except(exception_filter(GetExceptionCode())) {
> ```

Please use SIGACTION macro (ut_sigaction)

---

*[src/test/blk_recovery/blk_recovery.vcxproj, line 136 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOOnmX5jBjSHUYe-LWY:-KOOnmX5jBjSHUYe-LWZ:b-sppp5q) ([raw file](https://github.com/pmem/nvml/blob/88101937ff7e15d69241b82b7fa4847630e8971f/src/test/blk_recovery/blk_recovery.vcxproj#L136)):*
> ```XML
>       <CompileAs>CompileAsC</CompileAs>
>       <DisableSpecificWarnings>4996</DisableSpecificWarnings>
>       <ForcedIncludeFiles>platform.h;endian.h</ForcedIncludeFiles>
> ```

Please add endian.h as normal include

---

*[src/test/pmempool_info/TEST5.PS1, line 51 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOOo8PiNbHRSMzbbR5q:-KOOo8PjhXCK8p5tl5Ds:b-xtjv6f) ([raw file](https://github.com/pmem/nvml/blob/88101937ff7e15d69241b82b7fa4847630e8971f/src/test/pmempool_info/TEST5.PS1#L51)):*
> ```
> 
> require_fs_type any
> # pmemblk structure is different (i.e. bigger) in debug and our match
> ```

it should be XXX - (as we should print volatile part of pool header)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1057)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: 0 of 13 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/test/blk_recovery/blk_recovery.c, line 182 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOOnZD2ChlTPWYuLdeb:-KOSHAgXYFsBFBhebqBJ:b-7m5o8e) ([raw file](https://github.com/pmem/nvml/blob/88101937ff7e15d69241b82b7fa4847630e8971f/src/test/blk_recovery/blk_recovery.c#L182)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> Please use SIGACTION macro (ut_sigaction)
</details>
Thank you, fixed.

---

*[src/test/blk_recovery/blk_recovery.vcxproj, line 136 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOOnmX5jBjSHUYe-LWY:-KOPnSABChJMSX5CN8sM:b-ejh4vn) ([raw file](https://github.com/pmem/nvml/blob/88101937ff7e15d69241b82b7fa4847630e8971f/src/test/blk_recovery/blk_recovery.vcxproj#L136)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> Please add endian.h as normal include
</details>
I chose this because it should be working Linux without that. Included it in the file now. Thank you.

---

*[src/test/pmempool_info/TEST5.PS1, line 51 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOOo8PiNbHRSMzbbR5q:-KOPnp5K0PxsniJs5wrk:b-17884p) ([raw file](https://github.com/pmem/nvml/blob/88101937ff7e15d69241b82b7fa4847630e8971f/src/test/pmempool_info/TEST5.PS1#L51)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> it should be XXX - (as we should print volatile part of pool header)
</details>
If that's the thinking then let me fix it to print just the volatile part. By volatile, I assume you meant the non-debug part. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1057)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: 0 of 16 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/test/blk_recovery/blk_recovery.c, line 182 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOOnZD2ChlTPWYuLdeb:-KOf9pfuGUZfwtP6A_yE:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/88101937ff7e15d69241b82b7fa4847630e8971f/src/test/blk_recovery/blk_recovery.c#L182)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Thank you, fixed.
</details>
Done.

---

*[src/test/blk_recovery/blk_recovery.vcxproj, line 136 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOOnmX5jBjSHUYe-LWY:-KOf9pIVijijOdxdsTtO:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/88101937ff7e15d69241b82b7fa4847630e8971f/src/test/blk_recovery/blk_recovery.vcxproj#L136)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> I chose this because it should be working Linux without that. Included it in the file now. Thank you.
</details>
Done.

---

*[src/test/pmempool_info/TEST5.PS1, line 51 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOOo8PiNbHRSMzbbR5q:-KOf9osHd9S-meXjabrU:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/88101937ff7e15d69241b82b7fa4847630e8971f/src/test/pmempool_info/TEST5.PS1#L51)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> If that's the thinking then let me fix it to print just the volatile part. By volatile, I assume you meant the non-debug part. Fixed.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1057)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 7 of 11 files at r1, 10 of 10 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[src/test/RUNTESTS.ps1, line 119 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOiVkQeDaWmCXWGm7jq:-KOiVkQeDaWmCXWGm7jr:bdnoi4l) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/test/RUNTESTS.ps1#L119)):*
> ```PowerShell
> # get_build_dir -- returns the directory to pick the test binaries from
> #
> # exmaple, to get release build dir
> ```

example

---

*[src/test/blk_recovery/TEST0.PS1, line 66 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOiUbDn8EAG_Wj_jeFx:-KOiUbDn8EAG_Wj_jeFy:b2nh8oq) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/test/blk_recovery/TEST0.PS1#L66)):*
> ```
> # that testfile1 is consistent (after recovery steps were taken).
> #
> # expect_normal_exit $Env:EXE_DIR\blk_recovery$EXESUFFIX 4096 $DIR\testfile1 5 10
> ```

???

---

*[src/test/blk_recovery/TEST0.PS1, line 72 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOiUeRFQ2rFJTndKMLw:-KOiUeRFQ2rFJTndKMLx:b-of7cz6) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/test/blk_recovery/TEST0.PS1#L72)):*
> ```
> 
> check
> ```

missing "pass"

---

*[src/windows/win_signal.c, line 78 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOiufkSMgzn7hWtYcea:-KOiufkSMgzn7hWtYceb:b-amf5ai) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/windows/win_signal.c#L78)):*
> ```C
> /*
>  * string constants for strsignal
>  * XXX: ideally this should have the signal number as the suffix but then we
> ```

strsignal() is not guaranteed to be thread-safe, so you don't even need a TLS.
Linux implementation actually uses TLS, but if it fails for some reason, it falls back to static buffer.

Anyway, I think this is fine for now, as we only use it in our unit tests.

---

*[src/windows/win_signal.c, line 86 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOixJs20YfaQQBO_80Y:-KOixJs3ey2oFq_BrW6z:b-oar4f8) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/windows/win_signal.c#L86)):*
> ```C
> 
> /*
>  * strsignal(int sig)
> ```

strsignal -- return a string describing the meaning of the signal number 'sig'

---

*[src/windows/win_signal.c, line 91 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOixaQQdXX6LaeXWY-T:-KOixaQQdXX6LaeXWY-U:bqcah2g) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/windows/win_signal.c#L91)):*
> ```C
> strsignal(int sig)
> {
> 	if (sig < ARRAYSIZE(sys_siglist))
> ```

what if it's negative ? ;-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1057)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[src/windows/win_signal.c, line 78 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOiufkSMgzn7hWtYcea:-KOjW6PJ6MnWbHAjGn7V:b-640944) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/windows/win_signal.c#L78)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> strsignal() is not guaranteed to be thread-safe, so you don't even need a TLS.
> Linux implementation actually uses TLS, but if it fails for some reason, it falls back to static buffer.
> 
> Anyway, I think this is fine for now, as we only use it in our unit tests.
> 
</details>
Note: this is all new uncovered code, should add test coverage either here or as your next pull request

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1057)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: 13 of 16 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/test/RUNTESTS.ps1, line 119 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOiVkQeDaWmCXWGm7jq:-KOkAfGqt9gwYkya4RMy:bdbxn9h) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/test/RUNTESTS.ps1#L119)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> example
</details>
:) Done. Thank you.

---

*[src/test/blk_recovery/TEST0.PS1, line 66 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOiUbDn8EAG_Wj_jeFx:-KOkBE8iNfyXwjWoHk36:bv9kcxv) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/test/blk_recovery/TEST0.PS1#L66)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> ???
</details>
I'm equally puzzled. Fixed that now. Thank you.

---

*[src/test/blk_recovery/TEST0.PS1, line 72 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOiUeRFQ2rFJTndKMLw:-KOkBntOeadCJtr1p1Ip:bt2vr1b) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/test/blk_recovery/TEST0.PS1#L72)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> missing "pass"
</details>
Fixed.

---

*[src/windows/win_signal.c, line 78 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOiufkSMgzn7hWtYcea:-KOkCWBqs095JC0yFnOE:b-hwk18i) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/windows/win_signal.c#L78)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> Note: this is all new uncovered code, should add test coverage either here or as your next pull request
</details>
@krzycz: I see, thanks for letting me know, I added a note along with this comment.
@peluse: Make sense, I have added a trello card against me, will do that as a followup.

---

*[src/windows/win_signal.c, line 86 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOixJs20YfaQQBO_80Y:-KOkDjjw9ImPYW-JFhVv:b665541) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/windows/win_signal.c#L86)):*
> -- return a string describing the meaning of the signal number 'sig'
Fixed

---

*[src/windows/win_signal.c, line 91 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1057#-KOixaQQdXX6LaeXWY-T:-KOkDvd9EgTFDwAGB8D-:b-4mvhg6) ([raw file](https://github.com/pmem/nvml/blob/07b47e0a340c6f1a295b7460e25571494e98a6b6/src/windows/win_signal.c#L91)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> what if it's negative ? ;-)
</details>
Overlooked, thank you. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1057)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1057#-:-KOkHvIO3CAFZ7mPb3kb:bnfp4nl)*
<!-- Sent from Reviewable.io -->
